### ISSUE TITLE
Cannot login to private registry

### ISSUE THREAD

**Issue by ankiva** [docker/for-mac#5546]
Docker login fails with unresolved name.

docker login -u username registry-url
Password: 
Error response from daemon: Get https://registry-url.local/v2/: Bad Gateway

Logging in to hub succeeds.
From Console:
2021-04-06 11:02:59.409332+0300  localhost com.docker.backend[23430]: unexpected error from dnssd: Timeout (-65568)

and

2021-04-06 10:19:02.826740+0300  localhost kernel[0]: (Sandbox) System Policy: com.docker.backe(22788) deny(1) system-privilege 10006

  - [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID:7D03CBDD-9F6B-4495-91A3-7149FC2F2BD2/20210406073038

### Expected behavior
I can log in to private docker registry, name is resolved.
### Actual behavior
Resolving private registry name fails and login fails.
### Information

  - Is it reproducible? every time.
  - Is the problem new? happens for Apple M1, not encountered for Mac amd64
  - macOS Version: Big Sur 11.2.3 on Apple M1
  - Docker Desktop Version: Engine Version: 20.10.5

### Steps to reproduce the behavior

  1. open docker
  2. docker login private-registry

**Comment by djs55** [comment]
@ankiva thanks for your report. I think you're correct, it's a problem with the DNS resolution. I notice from the diagnostics that you're using a `.local` address (I assume mDNS). I tried this locally and it seems to work for me on my local network. As an experiment, could you try adding an explicit entry to `/etc/hosts` entry on the host to see if that works any better?

**Comment by ankiva** [comment]
I have dns configured in host's resolv.conf and the dns is meant to resolve the domain name, no mDNS. Pinging private servers with .local suffix in daemon vm fails, pinging servers with other suffix succeeds.
Adding explicit entries to /etc/hosts did not make any difference, still unresolvable.
Diagnostics ID: 7D03CBDD-9F6B-4495-91A3-7149FC2F2BD2/20210406100729

**Comment by jan-wilms** [comment]
We can reproduce this problem with an Intel x64 Mac.

**Comment by baronfel** [comment]
This hit me and my team at work today with our internal gitlab hosted on a `.local` address.  Rolling back from 3.3 worked for us, however the /etc/hosts file fix did not.  It's made doubly-frustrating by the fact that I cannot control the update check on the local client without a paid team plan.

**Comment by ankiva** [comment]
Did you flush cache also after the file update?

**Comment by baronfel** [comment]
@ankiva yup! I've got a little helper shell function that I use for it:

```sh
flushdns () {
	sudo killall -HUP mDNSResponder
	sudo killall mDNSResponderHelper
	sudo dscacheutil -flushcache
}
```

**Comment by thirdgen88** [comment]
I've also had this strike my team at work.  Unfortunately, like @baronfel, adding entries to `/etc/hosts` on the host didn't work for me either.  Hopefully this regression can be resolved soon!

**Comment by MarcoMatarazzo** [comment]
Same problem here, MacBook Pro 13 2019 on Mojave.

**Comment by ankiva** [comment]
@jan-wilms the behaviour now reproduces on our Intel Macs also starting from version Docker version 20.10.5, build 55c4c88. Workaround fixes the issue.

**Comment by baronfel** [comment]
Happy to report that docker for mac version 3.3.3 fixed this issue for me. Judging from the changelog the fix actually came with 3.3.2, but I had been delaying updates and so only saw the benefit when I updated 3.3.3.

**Comment by djs55** [comment]
@baronfel thanks for confirming the fix in 3.3.2 (and 3.3.3!) works. I'll close this issue now but feel free to open a fresh one if you hit further problems.
